### ISSUE TITLE
Add Progress Bar to Factory Production UI

### ISSUE THREAD

**Issue by ldicker83** [ldicker83/OPHD#34]
This is a lower priority but will help users to quickly visualize where a factory is at in terms of its current production whenever they bring up the Factory Production UI.

**Comment by ldicker83** [comment]
This should be a simple modification of the `FactoryProduction` window.

**Comment by DanRStevens** [comment]
In addition, I was thinking we should probably have a `ProgressBar` class derived from `Control`, rather than just a standalone method to draw progress bars given all the individual values.

It shouldn't be very difficult to convert the freestanding method to the `update` method of a new `Control` derived class. Though it might be wise to use integer values for the current progress value and the max progress value. Most cases where the progress bar are used have natural `int` inputs, so conversion to `float` is typically unnecessary.

**Comment by ldicker83** [comment]
Generally agreed. The stand alone functions were a quick and dirty type of deal which continue to be useful after implementation. :/

**Comment by belgianguy** [comment]
I tried to wrap my head around what was being discussed here, but I reckon it's about this UI Window:

Empty:
![image](https://user-images.githubusercontent.com/3091512/109370271-ed38fc00-789f-11eb-8735-9680fbac176e.png)

Item selected:
![image](https://user-images.githubusercontent.com/3091512/109370309-0e99e800-78a0-11eb-81fa-0050b716ddc8.png)

The Metal/Mineral values are the result of multiplying the number of turns and the amound of each mineral/metal they consume each turn.

This is to become graphical, like for example the MineReport (which has both a small and a big version):

![image](https://user-images.githubusercontent.com/3091512/109370485-d1822580-78a0-11eb-935e-aa7c9763bab1.png)

I reckon the general layout of the top 4 progress bars would be fitting for the tooltip, as the text is layed out the same way there.

I suppose the standalone functions that are being referred to is one in Common, called drawBasicProgressBar:

```
void drawBasicProgressBar(int x, int y, int width, int height, float percent, int padding)
{
	auto& renderer = Utility<Renderer>::get();
	renderer.drawBox(NAS2D::Rectangle{x, y, width, height}, NAS2D::Color{0, 185, 0});

	if (percent > 0.0f)
	{
		int bar_width = static_cast<int>(static_cast<float>(width - (padding + padding)) * percent);
		renderer.drawBoxFilled(NAS2D::Rectangle{x + padding, y + padding + 1, bar_width - 1, height - (padding + padding) - 1}, NAS2D::Color{0, 100, 0});
	}
}
```
Ideally, this is to be promoted to its own Control derived ProgressBar.


As a starting point, to see what an actually visible Control derived class looks like, I was looking at Label. But that derives from TextControl, so might not be the best suited, although its readonly nature does lend some inspiration I think, but rather than drawing text, you'd be drawing the progress bar. And as ProgressBar doesn't have any textual properties, it can probably just derive from Control.


I hope I got the gist of it. I'm very new to all of it, but this gives some dots to connect things to.

**Comment by belgianguy** [comment]
I tried something, but as it's all rather new to me, it's probably wise to wait for some feedback:

https://github.com/OutpostUniverse/OPHD/compare/master...belgianguy:AddProgressBarControl

![truck_progress](https://user-images.githubusercontent.com/3091512/109374422-80306100-78b5-11eb-9cdb-11c5e8f737a3.png)

![dozer_progress](https://user-images.githubusercontent.com/3091512/109374424-80c8f780-78b5-11eb-8b20-866f6857deb5.png)

Eventually I would place it where now the text "x of y turns completed" resides.

**Comment by DanRStevens** [comment]
I think you've got the gist of converting the progress bar code to a new `Control` derived class. That alone could be its own PR.

I'm less certain about which UI Leeor was looking to update. My initial thought was something involving the Factory Reports view, though you did indeed find a window with the exact title of "Factory Production".

I think the UI update could be done as a second PR. You could potentially develop both in parallel though, as a way of testing the new `Control` derived class. I would probably try to commit and push the changes separately though. Keeps the review chunks smaller, and makes it less likely that code will be hung up during review due to changes.

**Comment by DanRStevens** [comment]
I took a quick peek at the code you linked. Looks like the code has the right basic idea.

The NAS2D library has a `Vector` type, which can be used for `Control` `size` values. Additionally, the `Control` area is typically stored at the `mRect` field. A `Point` and `Vector` can be used to set a `Rectangle` object's position and size. You probably don't need to store explicit `mWidth` and `mHeight` fields in derived classes, as the `mRect` field in the base `Control` class should be able to represent those values. That may also simplify some of the draw calls, as you can pass the already constructed `mRect` value, rather than trying to reconstruct a new `Rectangle` object.

I think I would consider having a maximum value set in the `ProgressBar` constructor, and then be able to `set` the current value. Let the control take care of things like "percentage" calculation. Though internally it probably wouldn't use percent. Instead you could do things like:
```cpp
const auto barWidth = mRect.width() * mValue / mMaxValue;
```

The `ProgressBar` control has a border, and an inner draw area. The `mRect` value will probably represent the outer border bounds. You can use `Rectangle::inset` with the border thickness on `mRect` to convert to the inner draw area.

**Comment by ldicker83** [comment]
You've got the general idea, and yes, that was the UI window I was thinking of here.

I like Dan's idea of having the `ProgressBar` control taking a 'max count' type of value and then setting said value. It would do the 'percentages' on its own and sort of abstract that away from the user. As I mentioned in a previous comment, the `drawBasicProgressBar()` function was a 'quick and dirty' solution to just get something working. I really should have just spent the time making it a proper GUI Control but now's as good a time as any to start!

As for being new to the code base... yeah, I get that. I know what it's like exploring a body of work you've never seen before. NAS2D is the middleware that OutpostHD is built on -- they're both my projects but NAS2D has had considerable progress made with Dan and Brett's help (and also a user cugone but they've moved on to other projects).

Anyway, by all means ask away whenever there are questions. If you use Discord you can message me directly as well. Based on your username I assume we're on opposite sides of the planet so there may be some delay but I'm usually pretty responsive.

**Comment by ldicker83** [comment]
---

First suggestion, you can remove the header text:

```cpp
// ==================================================================================
// = NAS2D+GUI
// = Copyright © 2008 - 2020, Leeor Dicker
// ==================================================================================
// = This file is part of the NAS2D+GUI library.
// ==================================================================================
```

It's old and doesn't really add any value. It's also not actually part of the NAS2D library anymore. Not yet anyway! :D I've been meaning to get around to removing that from all of the UI controls.

**Comment by belgianguy** [comment]
Hi, Leeor and Dan, thanks for the guidance so far, I'm still working on my current branch, but once I have a more stable understanding I'll extract the ProgressBar control to its own branch and that will be in the lead. I just find it hard to develop the control without also checking that it fits it purpose and looks the part.

So far:
* The NAS2D comment has been removed
* The mWidth and mHeight members were removed and instead the mRect.width and mRect.height member variables are now used.
* Yes I reside in the country that is in my username
* I don't yet understand the maxValue approach, currenly this value gets set when the FactoryProduction constructor is called.
E.g. we know for example that a Truck takes 3 turns and a Dozer/Miner/Digger takes 5 turns to complete. How would we express that in relation to the maximum? Would we need to use the maxValue in the mValue calculation? Is a guard needed to prevent mValue from exceeding mMaxValue?
* I still need to investigate what the usage of Rectange::inset is.

**Comment by DanRStevens** [comment]
The `maxValue` will depend on what the progress bar represents. If it represents turns to completion, and a truck takes 3 turns, then `maxValue` will be 3, while the `value` will be incremented by 1 each turn. If a robot takes 5 turns to complete the `maxValue` will be 5. It would make sense to default the initial `value` to `0` in these cases. That's probably a sensible default in general. I don't know if need to provide an option to set `value` as a constructor parameter, though I'm open to allowing that.

The `Rectangle::inset` method adjusts corner points inwards by a certain amount. It can be used to adjust a border area to an inner contained area, given the `borderThickness`:
```cpp
Rectangle borderRect{0, 0, 10, 20};
Rectangle innerArea{borderThickness, borderThickness, 10 - borderThickness * 2, 20 - borderThickness * 2};

ASSERT_EQ(borderRect.inset(borderThickness), innerArea);
```

**Comment by belgianguy** [comment]
Hi Dan, the ProgressBar constructor is currently called when there is no concept yet of the product that is being built, so the maxValue (e.g.: 3 or 5) is not known at the point of creation. So for my current setup I'm scratching my head, or I am still missing something? Should there perhaps also be a setMaxValue setter that must be called whenever another product is selected from the grid (none vs miner vs truck)?


Current constructor:

```
ProgressBar::ProgressBar(int value, int maxValue, int padding)
{
	mValue = value;
	mMaxValue = maxValue;
	mPadding = padding;
}
```

private member in FactoryProduction

```
ProgressBar prbProduction{0, 5, 1};
```

**Comment by DanRStevens** [comment]
It would be reasonable to allow changing the `maxValue` with a setter. Just keep in mind that `maxValue` should probably never be `0`, so as to avoid division by 0 problems. You may also need to re-clamp the `value` if `maxValue` changes.

Not sure if setting `0` for `maxValue` should result in an immediate exception, or if the progress should just silently accept this and avoid division in that case, perhaps by forcing it to display as either 0% or 100%. I think I'm leaning towards immediate exception, as that would more likely allow programmer mistakes to be caught early.

----

If changing the selected item changes the `maxValue`, you could potentially re-construct the `ProgressBar` object and assign it over top of the old one. That would be one way to update the `maxValue` without needing a setter. I find this pattern useful when changing one aspect of an object requires changing most other aspects. Here's the main properties are `maxValue`, `value`, and the border thickness. If you changed `maxValue`, you would likely also be changing `value` along with it.

```cpp
prbProduction = ProgressBar{0, selection.turnsToCompletion, 1};
```

**Comment by belgianguy** [comment]
Thanks for that Dan, I also start to realize that a progress bar only makes sense as soon as a product is being built.

As a factory is created, it is not producing anything, so the first time the user opens the FactoryProduction window, there is no product being built, and there is no concept of max turns needed, thus there cannot and should not be any progress bar. So that means a progress bar can only be shown as soon as the user has selected a first item to produce. Meaning that the existance and visibility of the progress bar is dependent on the state of the Factory, here, having a product selected for production.

However, it in unclear to me if there ever then comes a scenario again in which the progressbar must be hidden again.
From a user experience point of view, it was not immediately clear what the function of the "Clear selection", "Apply", "Okay" and "Cancel" buttons all do. So it's also unclear what their effect on the progress bar (if any) should be.

**Comment by ldicker83** [comment]
Resetting the product will lose any resources spent on the previous product. It's a sort of a punishment for mismanagement to the user.

When the user switches it should reset everything back to 0.

---

Feedback about the interface is helpful. Clear Selection means pretty much that, set the factory to produce nothing. Apply applies any changes made without closing the window and Okay applies changes and closes the window. Admittedly this could be better so your thoughts on a better arrangement or better verbiage is helpful.

**Comment by Brett208** [comment]
@ldicker83, I have found the factory production UI confusing as well. It was helpful to read the description you just made of the buttons. Glad to hear it is getting addressed.

**Comment by DanRStevens** [comment]
Yeah, come to think of it, usage is not super clear to me either, particularly before reading that description.

I think what adds to the confusion is you can select a new product while an old one is being built. That leads to confusion over whether the progress display is for an old item, which isn't indicated anywhere what it is, versus a new item. Maybe the selection list could be hidden or locked while an item is being produced? You would need to cancel or complete the item before it is unlocked.

I'm not particularly fond of apply and cancel buttons. I figure buttons should have immediate effect. Particularly with my suggestion about locking the build selection while the factory is producing. I would expect buttons for either "Cancel Build" if one was in progress, or an unlocked selection list, and after selecting an item, a "Start Build" button would be enabled. Once a build is started, the progress bar would then display.

Alternatively,  you could have a split display where there is a selection, with stats about the selected item, and a build area, with stats about the item being built. The two need not coincide. I'm somewhat less fond of this approach though, as it presents more data on screen at once, and so more chance to confuse the user.

**Comment by ldicker83** [comment]
All good feedback.

That window is really kind of a hold-over from early development. It hasn't changed at all in literally years. It may make sense to redesign it from the ground up since it was essentially a quick and dirty "Let's get this working" kind of effort.

**Comment by belgianguy** [comment]
Okay, so we're diverging a bit here:
* I'll add the ProgressBar.h and ProgressBar.cpp files to their own branch.

Some other things I found out about the factory: 
* The resetting of turns still must be added when changing selection to another product. This is the punishment if I understand correctly.
* The grid selection change event is also fired if your current selected item is the already selected item. E.g. you are producing Truck and you select Truck again. This should be caught and do nothing, as that is the already selected product.
* When selecting another product to produce but the turns completed is larger than zero, it's perhaps nice to alert the user that he/she will waste resources when doing so. Is resetting the turns enough punishment for the user, and is that logically sound for the game state or does the production of the previous item explicitly be canceled?
* Selecting "No" at the above prompt did present another UI debacle, the selection would still change in the icon grid, even though the operation was canceled. I suggest we here also make sure the icongrid re-selects whatever it was producing before, as the user signalled that he/she changed his/her mind.
* Is there a "disabled" icon grid layout? e.g. "greyed out" icon grid and a "Forbidden" sign as cursor when in the icon grid, and unhooking all of its callbacks?
* Is there a NAS2D equivalent of a table structure similar to the one in HTML? It could allow for layouts that are nicely aligned without drawing them ourselves. Or is there another UI Control that can achieve this? 

Doodling around:
![prd_changed](https://user-images.githubusercontent.com/3091512/110179629-b1a2b280-7e08-11eb-8c90-fd55ddf815f4.png)

**Comment by Brett208** [comment]
@belgianguy, There is a StringTable that will allow laying out 2D grids of text. We have discussed expanding this to a more generic layout structure that could accept arbitrary UI elements, but I don't think it it has happened yet. There is also an IconGrid in OPHD, but I am less familiar with it although it sounds like you are already familiar with it.

For StringTable, see: https://github.com/OutpostUniverse/OPHD/blob/e8323144561889f63668ef5f5448a33f76eead60/OPHD/UI/StringTable.h

**Comment by belgianguy** [comment]
Hi @Brett208, the StringTable construction is exactly the one I'm trying to expand/replace. As I created the ProgressBar, I was looking on how I would add it to the right of the "Turns Completed:" text, but found no matching UI Component just yet. 

Thus a Table construct that can manage child Controls does not yet exist. :) I'll need to dig a little deeper I guess. Perhaps I can have a look if I can perhaps create something usable based on StringTable.

**Comment by ldicker83** [comment]
> * The grid selection change event is also fired if your current selected item is the already selected item. E.g. you are producing Truck and you select Truck again. This should be caught and do nothing, as that is the already selected product.

This is a good point. This should be a separate change set in the IconGrid control

> * When selecting another product to produce but the turns completed is larger than zero, it's perhaps nice to alert the user that he/she will waste resources when doing so. Is resetting the turns enough punishment for the user, and is that logically sound for the game state or does the production of the previous item explicitly be canceled?

I tend to dislike message boxes unless they're really used judiciously. Something like this, perhaps it makes sense to have that message box? I'll rely on the opinions of others to be sure.

> * Selecting "No" at the above prompt did present another UI debacle, the selection would still change in the icon grid, even though the operation was canceled. I suggest we here also make sure the icongrid re-selects whatever it was producing before, as the user signalled that he/she changed his/her mind.

I would make this part of the change set I mentioned above. This sounds like unintended behavior.

> * Is there a "disabled" icon grid layout? e.g. "greyed out" icon grid and a "Forbidden" sign as cursor when in the icon grid, and unhooking all of its callbacks?

There's a 'red' option -- see how it's handled with the structure icon grid on the main map view. If this isn't the behavior you're looking for that's definitely a feature that can be added. IconGrid, as it stands right now, is kind of a quick-and-dirty hacked together job that basically does the job but is anything but elegant. It could use some attention.

> * Is there a NAS2D equivalent of a table structure similar to the one in HTML? It could allow for layouts that are nicely aligned without drawing them ourselves. Or is there another UI Control that can achieve this?

I think @Brett208 mentioned the StringTable class which may be what you're looking for. I haven't done much with layout engines for the GUI... it's something @DanRStevens has been beating me over the head about and, while he's right, I find it tedious and have been dragging my feet on it. :/

> * Is the assumption correct that a product can only be changed when it is at 0 turns without penalty? Or are there other situations that would also allow this without penalty?

That's kind of what I was going for but I also want to have a 'notification' for when factories finish their production. Instead of every factory pushing a notification it would just be something like "One or more factories have completed their production cycle." with a button or link or something to pull up the full-screen UI to the factory tab. Something along those lines. I'm very open to suggestion, though, on how any of this can or should be handled.

**Comment by belgianguy** [comment]
> This is a good point. This should be a separate change set in the IconGrid control as firing off that event doesn't really make too much sense. Granted the Win32 API does fire events like that for Combo Boxes even if you select the same item which is probably why I didn't account for this case but it seems to me that this is something that should really be handled by the control itself.

This is addressed in #821 

> I tend to dislike message boxes unless they're really used judiciously. Something like this, perhaps it makes sense to have that message box? I'll rely on the opinions of others to be sure.

I too don't like MessageBoxes and will only use them if nothing else can be used. I've seen one occasion where it really startled me, and that is when trying to build a "red" building. Maybe we can replace that?
1.) I don't know what everything costs, so I'm not inclined to try and build them, only to then realize that it will cause a popup. Can we here not use some sort of audio to signal that this is an impossible action? Some negative beep, or a EVA voice that says "Insufficient resources"? 
2.) Would it not make more sense to add tooltips to the buildings in the IconGrid, which then could tell you what resources it will cost you (4x icon + text + number), and perhaps print in red those those resources that you are lacking when a building cannot be built?
3.) Why red? Isn't the "greyed out" the more known UI paradigm to signal that something is not available right now?
4.) Why is the Recycling Facility never red?
5.) When building something in a factory, but having no place to put it (no free slots and no warehouse) can we print the IDLE_REASON somewhere in the factory production window? Perhaps add a yellow footer with an triangle with an exclamation sign in the footer as to why this building is not producing? I picked yellow as red on dark grey is not as visible.

**Comment by DanRStevens** [comment]
These are all very good points. I particularly like the idea of using sound to highlight constraints due to resource shortages. And that combines well with tooltips and highlighted text in case sound is disabled (or the user is deaf).

**Comment by ldicker83** [comment]
> 1.) I don't know what everything costs, so I'm not inclined to try and build them, only to then realize that it will cause a popup. Can we here not use some sort of audio to signal that this is an impossible action? Some negative beep, or a EVA voice that says "Insufficient resources"?

This was a holdover until proper tooltips could be implemented.

> 2.) Would it not make more sense to add tooltips to the buildings in the IconGrid, which then could tell you what resources it will cost you (4x icon + text + number), and perhaps print in red those those resources that you are lacking when a building cannot be built?

See above.

> 3.) Why red? Isn't the "greyed out" the more known UI paradigm to signal that something is not available right now?

Because it's not simply not available, it's not available due to a lack of resource.

> 4.) Why is the Recycling Facility never red?

It's a lot cheaper than other structures. That's something I need to balance better.

> 5.) When building something in a factory, but having no place to put it (no free slots and no warehouse) can we print the IDLE_REASON somewhere in the factory production window? Perhaps add a yellow footer with an triangle with an exclamation sign in the footer as to why this building is not producing? I picked yellow as red on dark grey is not as visible.

Yes. Two things I want to do here, use the Notifications interface I'm working on and have that more obvious in the Structure Inspector. It still needs a little tlc.
